specs/capri/Monitoring Design.md

```diff
@@ -0,0 +1,742 @@
+
+* IOPs
+
+* Bandwidth
+
+* Latency
+
+Metrics to be collected for disk:
+
+* IOPs
+
+* Bandwidth
+
+* Latency
+
+Metrics to be collected for storage pool:
+
+* IOPs
+
+* Bandwidth
+
+* Latency
+
+Metrics to be collected for filesystem (later):
+
+* IOPs
+
+* Bandwidth
+
+* Latency
+
+Each Metrics driver should implement recommended metrics. Each driver can have
+its own custom metrics (key value pairs?).
+
+Action Item: Investigate how to store and retrieve custom metrics from
+Prometheus.
+
+Reference: http://stor2rrd.com/support_matrix.htm
+
+Open Issue: Metrics should be collected for ports as well. However ports are not
+managed by OpenSDS currently. This will be addressed later after we’ve decided
+whether/how to manage ports or not.
+
+### Pool or Backend Level
+
+* Capacities reported by the storage backend will be collected as metrics and
+emitted in Gauge metric type. The following will be collected at the storage
+backend pool level:
+
+* Volume_stats_total_capacity** (backend level) - Capri
+
+* Volume_stats_free_capacity** (backend level) - Capri
+
+TODO: more pool level stats
+
+Stats to monitor the health of the storage backend
+
+* Total number of volumes, snapshots, attached volumes, etc.
+
+* Configuration data
+
+* Data protection plan
+
+### Volume Level
+
+* Capacity data can also be collected at the per-volume level:
+
+* Volume_size (volume level) - Capri
+
+* Snapshot_size (volume level)
+
+TODO: more volume level stats
+
+* Volume_available_capacity (volume level)
+
+* Volume_total_capacity (volume level)
+
+* Volume_used_capacity (volume level)
+
+Note: More with thin provisioning support.
+
+### Operation Level
+
+* Count(or rate) and time it takes to perform certain volume operation will be
+collected and [Histogram
+type](<https://prometheus.io/docs/practices/histograms/>) will be used for
+emitting these metrics. For block storage service, this includes create volume,
+delete volume, expand volume, attach volume, detach volume, create snapshot,
+delete snapshot, create volume from snapshot. These data can be collected by the
+volume controller. An example is as follows:
+
+storage_operation_duration_seconds { storage_backend = "poolA", operation_name =
+"volume_create" }
+
+* Count of volume operation failures will be collected and emitted in histogram
+type as well.
+
+storage_operation_errors_total { storage_backend = "poolA", operation_name =
+"volume_attach" }
+
+* Count(or rate) and time it takes to perform create replication, delete
+replication, failover replication.
+
+* Count(or rate) and time it takes to perform upload object, download object,
+migrate object, etc. (multicloud local backend)
+
+### Metrics APIs
+-------------------
+
+*Metrics API*
+
+1.  POST API to trigger a collection
+
+/v1/{tenantId}/metrics
+
+*JSON request body*
+```json
+{
+"instanceId" : "1",
+"metrics" : ["a","b"],
+}
+```
+
+1.  instanceId is the specific instance of an entity like a pool or a volume
```

Comment:
Could u describe here which fields are required and which are optional?

Reply:
fixed, optional fields are marked with  //Optional